Rendering summary

NthTensor opened the pull request:
This PR adds a summary of the bevy's rendering crates as of 0.13. It is a work in progress, and absolutely contains omissions and mistakes. Please let me know if you find any. Feedback on both content and presentation is welcome! 

The goal is to cover the most important aspects of `bevy_render`, `bevy_core_pipeline` and `bevy_pbr` in order. Each crate is discussed in isolation, using only concepts introduced by it's dependencies. When this is complete (hopefully well before 0.14), I'd like to split this content up between the bevy book and the module docs of the relevant crates.

[RENDERED](https://gist.github.com/NthTensor/9d73a5ed75109d036fceb235b8f536a5)

NthTensor commented:
Do docs editors have a preference on filled vs unfilled paragraphs?

alice-i-cecile commented:
Overall feelings: really useful, not particularly user-approachable. I'll chew on the best way to digest this information into a format that matches the rest of the book.

> Do docs editors have a preference on filled vs unfilled paragraphs?

Not currently no.

andriyDev commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+and the graphics hardware. Going from highest to lowest level:
+1. The `WebGPU` api provided by `wgpu`,
+2. The platform-specific adapter (`wgpu_hal::vulkan`, `wgpu_hal::gles`,
+  `wgpu_native` and the browser's own `WebGpu` adapter, etc.)m
```
Extra m

andriyDev commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+TODO: Provide a brief list of features that differ across platforms.
+
+The middle sections of this stack are represented in in `wgpu` by a
+[`wgpu::Adapter`] and [`wgpu::Device`]. The adapter provides access the platform
```
"The adapter provides access **to** the platform..."

andriyDev commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+[`wgpu::Adapter`] and [`wgpu::Device`]. The adapter provides access the platform
+specific `WebGPU` implementation (layers 2 and 3 of the stack). The device is a
+logical representation of layers 4 and 5 obtained from the `Adapter`, and is
+ultimately what allows us to is submit work to the GPU.
```
Extra "is"

andriyDev commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+`wgpu` is a low level graphics library, but because it is designed to work
+across a wide beret of platforms there are many layers of abstraction between it
+and the graphics hardware. Going from highest to lowest level:
+1. The `WebGPU` api provided by `wgpu`,
```
This is my personal preference but I think we should reverse this list. For one, it's written from highest to lowest level, but the numbers go from lowest to highest. Also the abstraction feels less "obvious" in this order. As in, I think it makes more sense to say "start from the physical GPU, and slowly make it less and less specific". Maybe that's just me.

If you change the, make sure to renumber the references to the layers in following paragraphs!

andriyDev commented:
Ahhhh so sorry for the spam, I intended to start a review! 🤦‍♂️

NthTensor commented:
> Ahhhh so sorry for the spam

I appreciate the help no matter the format!

andriyDev commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+built using the `bevy_render` toolbox. It doesn't touch the actual shading, but
+it does set up a common `RenderGraph` and a set of standard `RenderPhases`.
+
+The main thing `bevy_core_pipleline` does is add two new subgraphs to the render
```
"pipeline"

andriyDev reviewed (changes requested):
Ok I've now finished my first pass. I've of course ignored all the TODO sections.

NthTensor commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+TODO: Expand the explanation of shaders.
+
+#### Layouts and Bind Groups
+Buffers, Texture Views and Samplers (as well as arrays of the same) can be
```
I ended up just redrafting this section. Your feedback was super helpful here especially.

NthTensor commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+
+The full list is available [here](https://www.w3.org/TR/webgpu/#limits).
+
+## `bevy_ecs` Preliminaries
```
I'm going to leave this in for now, but I agree it should be moved or deleted eventually.

NthTensor commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,852 @@
+[`RenderGraphContext::run_sub_graph`]: https://docs.rs/bevy/latest/bevy/render/render_graph/struct.RenderGraphContext.html#method.run_sub_graph
+
+#### CameraDriverNode and ViewNode
+With knowledge of subgraphs and views under our belt, we can approach one of the
```
I'll make a note to come back to this.

NthTensor commented:
Thank you so much! I've addressed your feedback. Going to start filling in some more of the TODO sections, and making notes for new topics to cover for the 0.14 release.

SIGSTACKFAULT commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,849 @@
+
+#### Shaders
+Shaders are small highly-specialized programs that run on the GPU within the
+programmable stages of pipelines. They are written in a language `WGSL`.
```
"They are written in the [`WGSL`](https://www.w3.org/TR/WGSL/) language"

link to https://www.w3.org/TR/WGSL/ or maybe the book page on it

SIGSTACKFAULT commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,849 @@
+Communication between the CPU and the GPU is asyncronuous; the CPU dispatches
+commands to the GPU and must wait for the GPU to finish it's workload and
+transmit a response. We'll reffer to these as GPU commands to differentiate them
+from bevy ECS commands.
```
"bevy ECS commands" needs link. also Bevy should be capitalized.

SIGSTACKFAULT commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,849 @@
+### Commands and Passes
+Communication between the CPU and the GPU is asyncronuous; the CPU dispatches
+commands to the GPU and must wait for the GPU to finish it's workload and
+transmit a response. We'll reffer to these as GPU commands to differentiate them
```
put "GPU commands" in quotes or italics maybe.

SIGSTACKFAULT commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,849 @@
+[`TrackedRenderPass`]: https://docs.rs/bevy/latest/bevy/render/render_phase/struct.TrackedRenderPass.html
+
+#### Parallel Command Encoding
+As covered in the `wgpu` preliminary, rendering involves queuing render commands
```
add internal link back to the wgpu preliminary header

SIGSTACKFAULT commented on content/learn/book/graphics/rendering-internals/technical-summary.md:
```diff
@@ -0,0 +1,849 @@
+main app or the render sub app. It effectively locks both worlds, and is the
+only part of bevy that can bottleneck both game logic and rendering. Bevy takes
+great pains to keep the extract schedule as slim and efficient as possible, and
+users should attempt to keep systems the extract schedule small.
```
missing an "in"

SIGSTACKFAULT reviewed (changes requested):
Obligatory half of these are IMHO

TimJentzsch commented:
Added the adopt me label, as it has been stale for over a year
